cmd/mailbox-controller/controller.go

```diff
@@ -53,46 +53,46 @@ const mbwsNameIndexKey = "mbwsName"
-	workspaceScopedLister      tenancylisters.WorkspaceLister
-	workspaceScopedClient      tenancyclient.WorkspaceInterface
-	apiBindingClusterInterface apisclient.APIBindingClusterInterface
-	queue                      workqueue.RateLimitingInterface // of mailbox workspace Name
+	context                 context.Context
+	espwPath                string
+	syncTargetInformer      cache.SharedIndexInformer
+	syncTargetIndexer       cache.Indexer
+	workspaceScopedInformer cache.SharedIndexInformer
+	workspaceScopedLister   tenancylisters.WorkspaceLister
+	workspaceScopedClient   tenancyclient.WorkspaceInterface
+	kbSpaceRelation         kbuser.KubeBindSpaceRelation
+	queue                   workqueue.RateLimitingInterface // of mailbox workspace Name
 }
 
+var suffix uint64 = 142857
+
 // newMailboxController constructs a new mailbox controller.
 // syncTargetClusterPreInformer is a pre-informer for all the relevant
 // SyncTarget objects (not limited to one cluster).
 func newMailboxController(ctx context.Context,
 	espwPath string,
-	syncTargetClusterPreInformer edgev2alpha1informers.SyncTargetClusterInformer,
+	syncTargetPreInformer edgev2alpha1informers.SyncTargetInformer,
 	workspaceScopedPreInformer kcptenancyinformers.WorkspaceInformer,
 	workspaceScopedClient tenancyclient.WorkspaceInterface,
-	apiBindingClusterInterface apisclient.APIBindingClusterInterface,
+	kbSpaceRelation kbuser.KubeBindSpaceRelation,
 ) *mbCtl {
-	syncTargetClusterInformer := syncTargetClusterPreInformer.Informer()
-	syncTargetClusterInformer.AddIndexers(cache.Indexers{mbwsNameIndexKey: mbwsNameOfObj})
+	syncTargetInformer := syncTargetPreInformer.Informer()
 	workspacesInformer := workspaceScopedPreInformer.Informer()
 
 	ctl := &mbCtl{
-		context:                    ctx,
-		espwPath:                   espwPath,
-		syncTargetClusterInformer:  syncTargetClusterInformer,
-		syncTargetClusterLister:    syncTargetClusterPreInformer.Lister(),
-		syncTargetIndexer:          syncTargetClusterInformer.GetIndexer(),
-		workspaceScopedInformer:    workspacesInformer,
-		workspaceScopedLister:      workspaceScopedPreInformer.Lister(),
-		workspaceScopedClient:      workspaceScopedClient,
-		apiBindingClusterInterface: apiBindingClusterInterface,
-		queue:                      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "mailbox-controller"),
+		context:                 ctx,
+		espwPath:                espwPath,
+		workspaceScopedInformer: workspacesInformer,
+		workspaceScopedLister:   workspaceScopedPreInformer.Lister(),
+		workspaceScopedClient:   workspaceScopedClient,
+		kbSpaceRelation:         kbSpaceRelation,
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "mailbox-controller"),
 	}
+	syncTargetInformer.AddIndexers(cache.Indexers{mbwsNameIndexKey: ctl.mbwsNameOfObj})
+	ctl.syncTargetInformer = syncTargetInformer
+	ctl.syncTargetIndexer = syncTargetInformer.GetIndexer()
```

Comment:
Why change from setting these fields in the mbCtl constructor above?

Reply:
Please see the comment above.

Reply:
I do not understand the relationship. `syncTargetInformer` and `syncTargetInformer.GetIndexer()` are available during the construction of the `mbCtl`.

Reply:
@MikeSpreitzer I'm confused. Can I execute `GetIndexer()` before `AddIndexers`?

Reply:
yes

Reply:
Changed

Reply:
Indeed.